site/app/controllers/AuthenticationController.php

```diff
@@ -308,4 +309,200 @@ public function userSelection() {
+            if ($is_family_name && $is_given_name) {
+                return true;
+            }
+            return false;
+        }
+        elseif ($requirements['require_email'] === true) {
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * Returns true if the password is greater than or equal to 12 characters, and has the required characters
+     */
+    public function isGoodPassword(string $password): bool {
+        return strlen($password) >= 12 && $this->checkChars($password);
+    }
+
+    /**
+     * @return array<string>
+     */
+    public function getAcceptedEmails(): array {
+        $json = file_get_contents('/usr/local/submitty/site/accepted_emails.json');
+        if ($json === false) {
+            return ['' => true];
+        }
+
+        // Decode the JSON file
+        $json_data = json_decode($json, true);
+
+        // Check if the JSON was decoded successfully
+        if ($json_data === null) {
+            return ['' => true];
+        }
+        return $json_data;  
+    }
+
+    /**
+     * Checks if the email extension is in the accepted emails JSON file
+     */
+    public function isAcceptedEmail(string $email): bool {
+        $json_data = $this->getAcceptedEmails();
+        // Check if the file was read successfully
+        try {
+            $email_extension = explode('@', $email)[1];
+        }
+        catch (\Error $error) {
+            return false;
+        }
+        return in_array($email_extension, array_keys($json_data), true);
+    }
+
+    /**
+     * Handles the submission of the new account creation form
+     */
+    #[Route("/authentication/self_add_user")]
+    public function addNewUser(): RedirectResponse {
+        // Check if the user is already logged in, if yes, redirect to home or another appropriate page
+        if ($this->logged_in) {
+            return new RedirectResponse($this->core->buildUrl(['home']));
+        }
+
+        // Should never happen, however they can visit this URL manually, so this is to prevent unwanted account creation.
+        if (!$this->core->getConfig()->isUserCreateAccount()) {
+            $this->core->addErrorMessage('Users cannot create their own account, Please have your system administrator add you.');
+            return new RedirectResponse($this->core->buildUrl(['authentication', 'login']));
+        }
+
+        $user_id = $_POST['user_id'];
+        $email = $_POST['email'];
+        $password = $_POST['password'];
+        $confirm_password = $_POST['confirm_password'];
+
+        $queried_list = $this->core->getQueries()->getFullEmailList();
+
+        if (in_array($email, array_column($queried_list, 'user_email'), true)) {
+            $this->core->addErrorMessage('Email already exists');
+            return new RedirectResponse($this->core->buildUrl(['authentication', 'create_account']));
+        }
```

Comment:
It would probably be more efficient to make SQL do this in_array check rather than having the entire user ID/email list sent back if it technically isn't needed.